### PR TITLE
Add IScreen interface and ScreenFactory

### PR DESCRIPTION
**Pull request by pavelanihimovsky**

### CODE REVIEW

**Comment by pavelanihimovsky** [comment]
resolves #19

File: src/main/java/aquality/appium/mobile/screens/screenfactory/ScreenFactoryProvider.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    @Inject
+    public ScreenFactoryProvider(IApplicationProfile applicationProfile) {
+        this.applicationProfile = applicationProfile;
+    }
+
+    @Override
+    public IScreenFactory getScreenFactory() {
+        if (applicationProfile.getPlatformName() == PlatformName.ANDROID) {
+            return new AndroidScreenFactory();
```
Comment (mialeska) [inline src/main/java/aquality/appium/mobile/screens/screenfactory/ScreenFactoryProvider.java:19]: How about single-return via ternary operator?

File: src/main/java/aquality/appium/mobile/application/MobileModule.java [hunk 2]
```diff
@@ -22,5 +25,6 @@ protected void configure() {
         bind(ILocalServiceSettings.class).to(getLocalServiceSettingsImplementation()).in(Singleton.class);
         bind(IConfiguration.class).to(getConfigurationImplementation());
         bind(IElementFactory.class).to(getElementFactoryImplementation());
+        bind(IScreenFactoryProvider.class).to(ScreenFactoryProvider.class);
```
Comment (knysh) [inline src/main/java/aquality/appium/mobile/application/MobileModule.java:28]: based on our current structure you should add a method getScreenFactoryProviderImplementation() in some module.
it is done to have a possibility to override implemetation of IScreenFactoryProvider in DI container.

File: src/main/java/aquality/appium/mobile/screens/Screen.java [hunk 3]
```diff
@@ -82,6 +60,13 @@ public String getName() {
         return name;
     }
 
+    /**
+     * Size of the element described by screen locator.
```
Comment (knysh) [inline src/main/java/aquality/appium/mobile/screens/Screen.java:64]: I think you could remove summary because it already exist in interface

File: src/test/java/integration/ScreenFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    public void shouldThrowIllegalArgumentExceptionOnNotValidPackageNameWithScreensValue()
+    {
+        System.setProperty(PACKAGE_NAME_WITH_SCREENS_VARIABLE_NAME, "fake.package");
+        AqualityServices.initInjector(new MobileModule(AqualityServices::getApplication));
+        Assert.assertThrows(IllegalArgumentException.class, () -> AqualityServices.getScreenFactory().getScreen(ILoginScreen.class));
+    }
+
+    @AfterClass
+    public void afterAll() {
+        System.clearProperty(PLATFORM_NAME_VARIABLE_NAME);
```
Comment (mialeska) [inline src/test/java/integration/ScreenFactoryTest.java:52]: I've reviewed test run (https://dev.azure.com/aquality-automation/aquality-automation/_build/results?buildId=1047&view=results) and from error messages and logs can see that the wrong profile for web tests was used. So you've probably broken this with setting/resetting of respectice environment variables.
Please review this.

File: azure-pipelines.yml [hunk 5]
```diff
@@ -1,157 +1,155 @@
...
+steps:
+  - task: SonarCloudPrepare@1
+    displayName: 'Prepare SonarCloud analysis'
+    inputs:
+      SonarCloud: 'SonarCloud'
+      organization: 'aqualityautomation'
+      scannerMode: 'CLI'
+      configMode: 'file'
+      extraProperties: |
+          sonar.coverage.exclusions=**/**
```
Comment (mialeska) [inline azure-pipelines.yml:26]: do we really need to change the pipeline in this PR? How is this related to ScreenFactory?

Comment (pavelanihimovsky) [reply azure-pipelines.yml:26]: I just formatted it

File: src/main/java/aquality/appium/mobile/configuration/ApplicationProfile.java [hunk 6]
```diff
@@ -42,4 +42,9 @@ public URL getRemoteConnectionUrl() {
             throw new InvalidArgumentException(String.format("Address %1$s provided in configuration by key %2$s is not valid", address, key));
         }
     }
+
+    @Override
+    public String getPackageNameWithScreens() {
```
Comment (DmitryBogatko) [inline src/main/java/aquality/appium/mobile/configuration/ApplicationProfile.java:47]: maybe getScreensPackageName sounds better?

File: src/main/java/aquality/appium/mobile/screens/Screen.java [hunk 7]
```diff
@@ -24,47 +22,27 @@
...
-     * Return screen state for screen locator
-     *
-     * @param timeout timeout for action
-     * @return True - screen is opened,
-     * False - screen is not opened
-     */
-    public boolean isDisplayed(Duration timeout) {
-        return screenLabel.state().waitForDisplayed(timeout);
+    @SuppressWarnings("unchecked")
+    protected T getDriver() {
```
Comment (DmitryBogatko) [inline src/main/java/aquality/appium/mobile/screens/Screen.java:33]: do we really need to return driver directly from the Screen, not from the AqualityServices?

Comment (mialeska) [reply src/main/java/aquality/appium/mobile/screens/Screen.java:33]: Yes, because this is driver casted to it's actual type (IOSDriver or AndroidDriver), and they have different interfaces inside. So if we want to perform smth custom with the driver (like start an activity for android or send a command to simulator on iOS), we should be doing it at screenObjects level. The getter to casted driver is needed here for this purpose

Comment (pavelanihimovsky) [reply src/main/java/aquality/appium/mobile/screens/Screen.java:33]: Not sure. @mialeska what do you think? Btw, it is protected, not public.

File: src/test/java/integration/ScreenFactoryTest.java (hunk 4 above)
Comment (pavelanihimovsky) [reply src/test/java/integration/ScreenFactoryTest.java:52]: You know, it is quite weired that env variable, that cleared after tests can affect other tests. So, the problem probably not in these tests.

File: src/main/java/aquality/appium/mobile/configuration/ApplicationProfile.java (hunk 6 above)
Comment (pavelanihimovsky) [reply src/main/java/aquality/appium/mobile/configuration/ApplicationProfile.java:47]: Not sure. This method is just a getter of the property from JSON. It should be the same as property.

File: src/test/java/samples/android/web/AndroidWebTest.java [hunk 8]
```diff
@@ -2,21 +2,19 @@
...
-import org.testng.annotations.BeforeGroups;
-import org.testng.annotations.Listeners;
+import org.testng.annotations.*;
 import testreport.ScreenshotListener;
 
 @Listeners(ScreenshotListener.class)
 public abstract class AndroidWebTest {
 
-    @BeforeGroups("web")
+    @BeforeClass
```
Comment (mialeska) [inline src/test/java/samples/android/web/AndroidWebTest.java:11]: With this change, tests would take longer, because browser will be started and exited many times. What is the purpose of this change?

Comment (pavelanihimovsky) [reply src/test/java/samples/android/web/AndroidWebTest.java:11]: Groups don't work here, so I changed to  beforeclass

Comment (mialeska) [reply src/test/java/samples/android/web/AndroidWebTest.java:11]: sure they worked, 
https://dev.azure.com/aquality-automation/aquality-automation/_build/results?buildId=838&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=fbe05904-ea97-54f8-0033-906e96004c63
tests were completed faster recently...
